tests/query/processors/test_filter_in_select_optimizer.py

```diff
@@ -0,0 +1,377 @@
+import pytest
+
+from snuba.datasets.factory import get_dataset
+from snuba.query import SelectedExpression
+from snuba.query.conditions import binary_condition
+from snuba.query.expressions import Column, FunctionCall, Literal
+from snuba.query.logical import Query
+from snuba.query.mql.parser import parse_mql_query
+from snuba.query.processors.logical.filter_in_select_optimizer import (
+    FilterInSelectOptimizer,
+)
+
+""" CONFIG STUFF THAT DOESNT MATTER """
+
+generic_metrics = get_dataset(
+    "generic_metrics",
+)
+mql_context = {
+    "entity": "generic_metrics_distributions",
+    "start": "2023-11-23T18:30:00",
+    "end": "2023-11-23T22:30:00",
+    "rollup": {
+        "granularity": 60,
+        "interval": 60,
+        "with_totals": "False",
+        "orderby": None,
+    },
+    "scope": {
+        "org_ids": [1],
+        "project_ids": [11],
+        "use_case_id": "transactions",
+    },
+    "indexer_mappings": {
+        "d:transactions/duration@millisecond": 123456,
+        "status_code": 222222,
+        "transaction": 333333,
+    },
+    "limit": None,
+    "offset": None,
+}
+
+""" IMPORTANT STUFF """
+
+optimizer = FilterInSelectOptimizer()
+mql_test_cases: list[tuple[str, set[int] | None]] = []
+logical_test_cases: list[tuple[Query, set[int] | None]] = []
+
+
+@pytest.mark.parametrize(
+    "mql_query, expected_domain",
+    mql_test_cases,
+)
+def test_get_domain_of_mql(mql_query: str, expected_domain: set[int]) -> None:
+    logical_query, _ = parse_mql_query(str(mql_query), mql_context, generic_metrics)
+    assert isinstance(logical_query, Query)
+    res = optimizer.get_domain_of_query(logical_query)
+    assert res == expected_domain
+
+
+@pytest.mark.parametrize("logical_query, expected_domain", logical_test_cases)
+def test_get_domain_of_logical(logical_query: Query, expected_domain: set[int]):
+    assert optimizer.get_domain_of_query(logical_query) == expected_domain
+
+
+""" --- MQL TEST CASES --- """
+
+mql_queries = [
```

Comment:
Can we validate with a test that the expected results of a query is still the same, with or without the processor?

Reply:
I had the same idea but I couldnt figure out an easy way to do this. After asking Evan about it, I think he didnt think it would be necessary. As a safeguard against breaking existing queries I put the optimizer behind a feature flag.

Reply:
> What is the behaviour of this processor when a query's SELECT clause aggregates two different metric_ids separately? Can we write a test that captures that case?

You bring up a good point. Currently the optimizer assumes that if there is any conditional aggregate function in the query, it's domain is the domain of the entire query. 

Thus if I have a query like `SELECT sumIf(value, metric_id=10), sumIf(value, metric_id=11)`

It will compute the domain of the first sumIf as 10, be finished, and add `WHERE metric_id=10`
Thus the query will be broken.